src/services/auto-type.service.ts

```diff
@@ -0,0 +1,936 @@
+ */
+function beep(freq: number, duration: number) {
+    const context = new AudioContext();
+    const o = context.createOscillator();
+    const g = context.createGain();
+    o.connect(g);
+    g.connect(context.destination);
+    o.frequency.value = freq;
+    o.start();
+    o.stop(context.currentTime + (duration / 1000));
+}
+
+function matchToArgs(match: RegExpMatchArray): [string, ...string[]] {
+    const index = match.findIndex((val: string, i: number) => (i > 0 && val !== ''));
+    return [match[index], ...match.slice(index + 1)];
+}
+
+export class AutoTypeService implements AutoTypeServiceBase {
+    private readonly systemClipModifier: string;
+
+    constructor(private storageService: StorageService, private cipherService: CipherService,
+                private platformUtilsService: PlatformUtilsService, private messagingService: MessagingService,
+                private passwordGenerationService: PasswordGenerationService) {
+        this.systemClipModifier =
+            this.platformUtilsService.getDevice() === DeviceType.MacOsDesktop ? 'command' : 'ctrl';
+    }
+
+    init() {
+        app.on('will-quit', () => {
+            // Unregister all shortcuts.
+            globalShortcut.unregisterAll();
+        });
+        app.on('ready', () => {
+            // Register auto-type hotkey
+            this.update();
+        });
+    }
+
+    update() {
+        globalShortcut.unregisterAll();
+        this.storageService.get<string>(ElectronConstants.enableAutoTypeKey).then((enabled) => {
+            if (enabled) {
+                this.storageService.get<string>(ElectronConstants.AutoTypeHotkeyKey).then(
+                    (accelerator) => {
+                        if (accelerator) {
+                            globalShortcut.register(accelerator, () => {
+                                this.typeTarget();
+                            });
+
+                            if (!globalShortcut.isRegistered(accelerator)) {
+                                throw new Error('Failed to register auto-type hotkey.');
+                            }
+                        }
+                    });
+            }
+        });
+    }
+
+    async getTarget(): Promise<string> {
+        return new Promise((resolve, reject) => {
+            const timer = setTimeout(reject, 10000, 'Took too long to select a window');
+            windowManager.once('window-activated', (window: Window) => {
+                clearTimeout(timer);
+                resolve(window.getTitle());
+            });
+        });
+    }
+
+    getPossibleTargets(): string[] {
+        return windowManager.getWindows().map((w: Window) => w.getTitle());
+    }
+
+    typeUsingClip(str: string) {
+        if (str.length === 0) { return; }
+        this.platformUtilsService.copyToClipboard(str);
+        robot.keyTap('v', this.systemClipModifier);
+    }
+
+    async typeTarget() {
+        const target = await this.getTarget();
```

Comment:
Looks like I missed something while writing this. typeTarget() needs to take the target as a parameter. Line 237 needs to be changed to `typeTarget(windowManager.getActiveWindow())`.